content/post/2019-11-01-internal-functions.Rmarkdown

```diff
@@ -0,0 +1,185 @@
+
+### Where to put internal functions?
+
+You could save internal functions used in one function only in the R file defining that function, and internal functions used in several other functions in a single utils.R file or specialized utils-dates.R, utils-encoding.R files. Choose a system that helps you and your collaborators find the internal functions easily, R will never have trouble finding them. :wink:
+
+Another possible approach to helper functions when used in several packages is to pack them up in a package such as [Yihui Xie's `xfun`](https://github.com/yihui/xfun). So then they're no longer internal functions. :dizzy_face: 
+
+### How to document internal functions?
+
+You should at least add a few comments in their code as usual. Best practice recommended in the [tidyverse style guide](https://style.tidyverse.org/documentation.html#internal-functions) and the [rOpenSci dev guide](https://devguide.ropensci.org/building.html) is to document them with roxygen2 tags like other functions, but to use `#' @NoRd` to prevent manual pages to be created. 
+
+```r
+#' Compare x to 1
+#' @param x an integer
+#' @NoRd
+is_one <- function(x) {
+  x == 1
+}
+
+```
+
+The keyword `@keyword internal` would mean [a manual page is created but not present in the function index](https://roxygen2.r-lib.org/articles/rd.html#indexing). A confusing aspect is that you could use it for an *exported, not internal* function you don't want to be too visible, e.g. a function returning the default app for OAuth in a package wrapping a web API.
+
+```r
+#' A function rather aimed at developers
+#' @description A function that does blabla, blabla.
+#' @keywords internal
+#' @export
+does_thing <- function(){
+ message("I am an exported function")
+}
+```
+
+## Explore internal functions
+
+You might need to have a look at the guts of a package when wanting to contribute to it, or at the guts of several packages to get some inspiration for your code.
+
+### Explore internal functions within a package
+
+Say you've started working on a new-to-you package (or resumed work on a long forgotten package of yours :wink:). How to know how it all hangs together?
+
+One first way to understand what a given helper does is looking at its code, [from within RStudio there are some useful tools for navigating functions](https://support.rstudio.com/hc/en-us/articles/200710523-Navigating-Code). You can then search for occurrences of its names across R scripts. These first two tasks are static code analysis (well unless your brain really executes R code by reading it!). Furthermore, a non static way to explore a function use [`browser()` inside it or inside functions calling it](https://resources.rstudio.com/rstudio-conf-2019/box-plots-a-case-study-in-debugging-and-perseverance).
```

Comment:
"Furthermore, a non static way to explore a function use [`browser()`"

-> "is to use" or something like that.